### PR TITLE
Using mostly the HDF5 C++ API. All HDF5 I/O should now be exception-s…

### LINKED ISSUE #254: Makeover for HDF5 I/O
Several functions there are currently not exception-safe. When fixing this, HDF5 calls should be changed to use the C++ API.

### PR DESCRIPTION
**Pull request by dweindl**
…afe. Closes #254.

### CODE REVIEW

**Comment by dweindl** [comment]
@FFroehlich Works now with master. However, the error checking is not very strict, since our matlab-written data can be somewhat messy. For some functions we need to decide which fields are mandatory and which ones are optional. Otherwise some tests might silently be skipped. 
One issue is for example how to deal with empty attributes (when writing matlab `[]`). Should this be considered valid input?

**Comment by FFroehlich** [comment]
hdf5 fields should only be skipped if the result is empty. If the testing routine cannot read a specific field, it should check whether the respective field in rdata is empty and throw an error otherwise.

File: include/amici_hdf5.h [hunk 1]
```diff
@@ -2,28 +2,49 @@
...
 class Model;
 class Solver;
 
-/* Functions for reading and writing AMICI data to/from HDF5 files. */
+namespace hdf5 {
+
+
+/** @file Functions for reading and writing AMICI data to/from HDF5 files. */
+
 // TODO: proper type checking, exception instead of return code, c++ api; check if all fields are read and saved
```
Comment (FFroehlich) [inline include/amici_hdf5.h]: this is no longer TBD right?

Comment (FFroehlich) [reply include/amici_hdf5.h]: at least partially

Comment (dweindl) [reply include/amici_hdf5.h]: Partially. Let's leave it until we changed the rest and adapted to new storage format.

**Comment by FFroehlich** [comment]
but lets adress the skipping of fields in #265. some of the "mandatory" fields are not written and NaN values/empty is not handled consistently.

**Comment by dweindl** [comment]
Agreed.

**Pull request merged**